helm/resource_release_test.go

```diff
@@ -925,6 +1010,64 @@ func testAccHelmReleaseConfigEmptyVersion(resource, ns, name string) string {
 	`, resource, name, ns, testRepositoryURL)
 }
 
+func testAccHelmReleaseConfigWithUpgradeStrategy(resource, ns, name, version string, enabled, install bool) string {
+	return fmt.Sprintf(`
+		resource "helm_release" "%s" {
+ 			name        = %q
+			namespace   = %q
+			description = "Test"
+			repository  = "%s"
+  			chart       = "test-chart"
+			version     = %q
+
+			upgrade {
+				enable = %t
+				install = %t
+			}
+
+			set {
+				name = "foo"
+				value = "qux"
+			}
+
+			set {
+				name = "qux.bar"
+				value = 1
+			}
+
+			set {
+				name = "master.persistence.enabled"
+				value = false # persistent volumes are giving non-related issues when testing
+			}
+			set {
+				name = "replication.enabled"
+				value = false
+			}
+		}
+	`, resource, name, ns, testRepositoryURL, version, enabled, install)
+}
+
+func testAccHelmReleaseConfigWithUpgradeStrategyWarmstart(ns, name string) string {
+	return fmt.Sprintf(`
+		resource "helm_release" "test" {
+ 			name        = %q
+			namespace   = %q
+			description = "Test"
+  			chart       = "./test-chart-1.2.3.tgz"
+			version     = "0.1.0"
+
+			upgrade {
```

Comment:
after some further thinking on how this upgrade is implemented, I wonder why we a block is necessary in this case when a simpler method would be to include this as a bool field:
`upgrade_install = true`

instead of having to include a whole block:
```
			upgrade {
				enable = true
				install = false
			}
```

This comes from wanting to simplify this a bit more.

Reply:
I had honestly gone back and forth on this myself, and what I eventually came down on the side of was trying to map the schema here as closely as possible to the way the helm cil logic works: `helm uprade` is a different top-level command from `helm install`, and `helm upgrade --install` has behavior notably different than `helm upgrade` without the `--install` flag.

I also didn't want to presuppose how other people would use the feature — I can't imagine a scenario where I would want to enable upgrade without install=true, but it's definitely an approach supported and documented by the helm CLI so I wanted to make the provider schema map to that.

Reply:
Thanks for the clarification! Let's go ahead with applying this as a boolean field. I appreciate you thinking this through, in regards to wanting to support `helm upgrade` we won't have to worry too much about this since the version itself can be upgraded on its own by updating the `version` field. 

Only other reasons to include a standalone `helm upgrade` is:

1. have the ability to automatically perform an upgrade to the latest version which we would not want. (As stated in a previous comment:
>I also attempted the following which matches the `helm upgrade <RELEASE> <CHART>` which upgrades the chart to the latest version
>
>1. installed `bitnami/redis` with no version set, marking the version as the latest (`19.6.1`)
>2. explicitly add `version = 19.6.0`, changing the version `19.6.1` -> `19.6.0`
>3. remove `version` to see whether the version is upgrade to `19.6.1`This results in no changes outputted by >terraform
>
>This is good news since this PR should **only** support `helm upgrade --install`, auto upgarding to the latest release is not something we want to support in the provider. This is noted [here](https://github.com/hashicorp/terraform-provider-helm/issues/102#issuecomment-621311452)

2. Resolve the `Error: cannot re-use a name that is still in use`, though this should be resolved with `helm upgrade --install` (i.e having a boolean field `upgrade_install = true`

This will simplify it more and should lead to a smaller block of documentation necessary for this new field.